apis/v1alpha1/httproute_types.go

```diff
@@ -165,32 +180,68 @@ type HTTPRouteMatch struct {
+)
+
+// HTTPRouteFilter defines additional processing steps that must be completed
+// during the request or response lifecycle.
+// HTTPRouteFilters are meant as an extension point to express additional
+// processing that must be done in Gateway implementations. Some examples include
+// request or response modification, implementing authentication strategies,
+// rate-limiting, and traffic shaping.
+// API guarantee/conformance is defined based on the type of the filter.
+// +union
+type HTTPRouteFilter struct {
+	// Type identifies the filter to execute.
+	// Types are classified into two conformance-levels:
+	// - Core and extended: These filter types and their corresponding configuration
+	//   is defined in this package. All implementations must implement
+	//   the core filters. Implementers are encouraged to support extended filters.
+	//   Definitions for filter-specific configuration for these
+	//   filters is defined in this package.
+	// - Custom: These filters are defined and supported by specific vendors.
+	//   In future, filters showing convergence in behavior across multiple
+	//   implementations will be considered for inclusion in extended or core
+	//   conformance rings. Filter-specific configuration can be specified
+	//   using ExtensionRef or Config fields.
+	// Implementers are encouraged to define custom implementation
+	// types to extend the core API with implementation-specific behavior.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	Type string `json:"type" protobuf:"bytes,3,opt,name=type"`
 
 	// ExtensionRef is an optional, implementation-specific extension to the
 	// "filter" behavior.  The resource may be "configmap" (use the empty
 	// string for the group) or an implementation-defined resource (for
 	// example, resource "myroutefilters" in group "networking.acme.io").
 	// Omitting or specifying the empty string for both the resource and
-	// group indicates that the resource is "configmaps".  If the referent
-	// cannot be found, the "InvalidRoutes" status condition on any Gateway
-	// that includes the HTTPRoute will be true.
-	//
-	// Support: custom
-	//
+	// group indicates that the resource is "configmaps".
+	// ExtensionRef MUST NOT be used for core and extended filters.
 	// +optional
 	ExtensionRef *RouteFilterExtensionObjectReference `json:"extensionRef" protobuf:"bytes,2,opt,name=extensionRef"`
+
+	// Config allows for inline-definition of filtering semantics.
+	// It provides an alternative for cases where a CRD is an overkill.
+	// Config MUST NOT be used for core and extended filters.
+	// +optional
+	Config map[string]string `json:"config" protobuf:"bytes,4,rep,name=config"`
+
+	// Filter-specific configuration definitions for core and extended filters
+
+	HTTPRequestHeaderTransformer *HTTPRequestHeaderTransformerConfig `json:"requestHeaderTransformer" protobuf:"bytes,1,opt,name=requestHeaderTransformer"`
```

Comment:
Does this mean that we'd have fields for all core and extended filters embedded in this type? This seems like it could get quite messy though I also don't have a great alternative.

Reply:
Indeed. The only other option here is to not define struct types, which also mean we lose a lot of validation. I do agree that this can and certainly will get cumbersome.

Reply:
I don't see alternative since they are all "supported" and should have stable APIs.
We can work on making the naming more amendable to users though:

Maybe:

```
requestHeaders:
- add:
  foo: foo-v
- remove: ["internal-header1"]
```

Reply:
Updated the Filter name and config struct to `RequestHeader`.
As discussed in office hours yesterday, this is super ugly but there does not seem be a better way to accomplish this.
Resolving this thread, please re-open if needed.